keras_cv/layers/preprocessing3d/group_points_by_bounding_boxes.py

```diff
@@ -0,0 +1,198 @@
+            raise ValueError("max_points_per_bounding_boxes must be >=0.")
+        if min_points_per_bounding_boxes > max_points_per_bounding_boxes:
+            raise ValueError(
+                "max_paste_bounding_boxes must be >= min_points_per_bounding_boxes."
+            )
+
+        self._label_index = label_index
+        self._min_points_per_bounding_boxes = min_points_per_bounding_boxes
+        self._max_points_per_bounding_boxes = max_points_per_bounding_boxes
+        self._iou_3d = iou_3d.IoU3D
+
+    def augment_point_clouds_bounding_boxes(
+        self, point_clouds, bounding_boxes, **kwargs
+    ):
+        if self._label_index > 0:
+            bounding_boxes_mask = tf.math.equal(
+                bounding_boxes[0, :, 7], self._label_index
+            )
+            object_bounding_boxes = tf.boolean_mask(
+                bounding_boxes, bounding_boxes_mask, axis=1
+            )
+        else:
+            bounding_boxes_mask = ~tf.math.equal(bounding_boxes[0, :, 7], 0.0)
+            object_bounding_boxes = tf.boolean_mask(
+                bounding_boxes, bounding_boxes_mask, axis=1
+            )
+
+        points_in_bounding_boxes = is_within_box3d(
+            point_clouds, object_bounding_boxes[:, :, :7]
+        )
+        # Filter bounding boxes using the current frame.
+        min_points_filter = (
+            tf.reduce_sum(tf.cast(points_in_bounding_boxes[0], dtype=tf.int32), axis=0)
+            >= self._min_points_per_bounding_boxes
+        )
+
+        object_bounding_boxes = tf.boolean_mask(
+            object_bounding_boxes, min_points_filter, axis=1
+        )
+
+        points_in_bounding_boxes = tf.boolean_mask(
+            points_in_bounding_boxes, min_points_filter, axis=2
+        )
+        # [num of frames, num of boxes, num of points].
+        points_in_bounding_boxes = tf.transpose(points_in_bounding_boxes, [0, 2, 1])
+        points_in_bounding_boxes = tf.cast(points_in_bounding_boxes, tf.int32)
+        sort_valid_index = tf.argsort(
+            points_in_bounding_boxes, axis=-1, direction="DESCENDING"
+        )
+        sort_valid_mask = tf.gather(
+            points_in_bounding_boxes, sort_valid_index, axis=2, batch_dims=2
+        )[:, :, : self._max_points_per_bounding_boxes]
+        # [num of frames, num of boxes, self._max_points_per_bounding_boxes, num of point features].
+        object_point_clouds = point_clouds[:, tf.newaxis, :, :]
+        num_valid_bounding_boxes = tf.shape(object_bounding_boxes)[1]
+        object_point_clouds = tf.tile(
+            object_point_clouds, [1, num_valid_bounding_boxes, 1, 1]
+        )
+        object_point_clouds = tf.gather(
+            object_point_clouds, sort_valid_index, axis=2, batch_dims=2
+        )[:, :, : self._max_points_per_bounding_boxes, :]
+
+        object_point_clouds = tf.where(
+            sort_valid_mask[:, :, :, tf.newaxis] > 0, object_point_clouds, 0.0
+        )
+
+        return (
```

Comment:
following up on the bb vs object_bb -- the output of this layer, or the input to all following layers, should it be a dict, and should the keys be "point_cloud" and "bounging_boxes"?

Reply:
Yes, I think the output of this layer should be a dictionary of {'point_clouds': tf.Tensor, 'bounding_boxes': tf.Tensor, 'object_point_clouds': tf.Tensor, 'object_bounding_boxes': tf.Tensor}